tf2onnx/optimizer/transpose_optimizer.py

```diff
@@ -101,42 +102,50 @@ def post_optimize_action(self):
-            #     indice = graph.make_const(utils.make_name("indice"), np.array([0, 3, 1, 2])).output[0]
-            # else:
-            #     indice = graph.make_const(utils.make_name("indice"), np.array([0, 2, 3, 1])).output[0]
             perm = list(op.get_attr('perm').ints)
             indices = graph.make_const(utils.make_name("indices"), np.array(perm)).output[0]
 
             return graph.make_node("Gather", [input_shape, indices]).output[0]
 
+        def check_transpose_replaceable_with_reshape(shape, perm):
+            shuffled = []
+            shifted = []
+            index = 0
+            for i in range(len(perm)):
+                if perm[i] != index:
+                    shuffled.append(perm[i])
+                else:
+                    if index != i:
+                        shifted.append(perm[i])
+                    index += 1
+            return all([shape[i] == 1 for i in shuffled]) or all([shape[i] == 1 for i in shifted])
+
         nodes = self.nodes
-        # if channel==1 or height==width==1, replace transpose with reshape
+        # if all of input dimensions are not shuffled at all
+        # or if the shuffled dimensions are all of size 1
+        # or if the shifted dimensions are all of size 1
+        # then replace Transpose with a Reshape
         # replacing trans with reshape is because transpose will copy data even if this transpose doesn't nothing
-#         for op in nodes:
-#             if op.type == "Transpose":
-#                 input_shape = self._g.get_shape(op.input[0])
-#                 if not input_shape:
-#                     continue
-# # FIXME
-#                 if (is_nchw_transpose(op) and (input_shape[3] == 1 or (input_shape[1:3] == [1, 1])))\
-#                    or (is_nhwc_transpose(op) and (input_shape[1] == 1 or (input_shape[2:4] == [1, 1]))):
-#                     new_shape = _calculate_new_shape(self._g, op)
-#                     # replace transpose with reshape
-#                     self._g.remove_node(op.name)
-#                     self._g.make_node("Reshape", [op.input[0], new_shape], name=op.name, outputs=op.output)
-#                     self._g.topological_sort(self._g.get_nodes())
+        for op in nodes:
+            if op.type == "Transpose":
+                input_shape = self._g.get_shape(op.input[0])
+                perm = list(reversed(range(len(input_shape))))
+                if op.has_attr('perm'):
+                    perm = list(op.get_attr('perm').ints)
+                if not input_shape:
+                    continue
+                if check_transpose_replaceable_with_reshape(input_shape, perm):
+                    new_shape = _calculate_new_shape(self._g, op)
+                    # replace transpose with reshape
+                    self._g.remove_node(op.name)
+                    self._g.make_node("Reshape", [op.input[0], new_shape], name=op.name, outputs=op.output)
+                    self._g.topological_sort(self._g.get_nodes())

```

Comment:
wonder if this needs to be in the inner loop since we loop over nodes. 
I guess it's not going to happen a lot, maybe ok.